cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -621,11 +621,21 @@ class DatasetWriter::DatasetWriterImpl {
         backpressure = writer_state_.open_files_throttle.Acquire(1);
         if (!backpressure.is_finished()) {
           EVENT_ON_CURRENT_SPAN("DatasetWriter::Backpressure::TooManyOpenFiles");
+          writer_state_.rows_in_flight_throttle.Release(next_chunk->num_rows());
           RETURN_NOT_OK(TryCloseLargestFile());
           break;
         }
       }
-      RETURN_NOT_OK(dir_queue->StartWrite(next_chunk));
+      auto s = dir_queue->StartWrite(next_chunk);
+      if (!s.ok()) {
+        // If `StartWrite` succeeded, it will Release the
+        // `rows_in_flight_throttle` when the write task is finished.
+        //
+        // `open_files_throttle` will be handed by `DatasetWriterDirectoryQueue`
+        // so we don't need to release it here.
+        writer_state_.rows_in_flight_throttle.Release(next_chunk->num_rows());
```

Comment:
It seems that the added test is passed without this.
Is it difficult to write a test for this change?

Reply:
Aha the added test hang forever on my machine😂 let me take a look...

Reply:
I've update the test, and make it holding less time and easier to reproduce the problem. The test run in MacOS with RelWithDebInfo before this patch will hang forever

Reply:
Oh I got what you mean, sorry for that! I think I just test https://github.com/apache/arrow/pull/38885/files#diff-387ad04c2450a38044e667e07183b8265866cb3736d10acdce137c2b83737b16R624

`DatasetWriterDirectoryQueue::StartWrite` is a bit hard to test, because `OpenFileQueue` always return OK now, `DatasetWriterFileQueue::Push` also merely failed

Reply:
OK.